src/types.mligo

```diff
@@ -146,6 +152,16 @@ type governance_token =
   ; token_id : token_id
   }
 
+type quorum_threshold_at_stage =
+  { quorum_threshold : quorum_threshold
+  ; last_updated_period : nat
+  }
+
+type period_staked =
+  { last_updated_period : nat
+  ; staked : nat
+  }
```

Comment:
Can we not track these together?

AFAIU the quorum changes only per-cycle, so going from a `voting` stage to a `proposing` one, but not when going from a `proposing` to a `voting` one; and what is staked also needs to be tracked in a similar way, right?

If that's the case I propose using:
```suggestion
type quorum_threshold_at_stage =
  { quorum_threshold : quorum_threshold
  ; last_updated_stage : nat
  // ^ last stage number in which the quorum was updated
  ; total_staked_cycle : nat
  // ^ total number of tokens staked in the cycle (proposing + voting stage)
  // that the last_updated_stage is part of
  }
```

Reply:
>Can we not track these together?

@pasqu4le But then we will only have a single `last_updated_period`, right? But we need to track the last updated period for staked tokens and quorum threshold separately. 

Because for updating staked count, we should check if it was updated in older an older cycle (proposing + voting periods, and if yest, we ll have to overwrite what ever the current value was, or else we need to add to the current value.

Updating quorum_threhold should come in the next cycle. So if we use the last_updated_period, then we will only see the last period where propose or vote operations were called.

Reply:
> we need to track the last updated period for staked tokens and quorum threshold separately.

Maybe I'm missing something, but I'm not sure we need to, this looks like the same situation as [the one we discussed before](https://github.com/tqtezos/baseDAO/pull/225#discussion_r628260543) for `frozen_total_supply`/`freeze`, except now it's for `staked_total_supply`/staking.

What I propose is that we update this value in the storage every time we `propose`, `vote` (because they update the staked total) and `flush` (because that's where the quorum is needed).

Say in the `last_updated_stage` we can have both `voting` and `proposing` numbers, then as before all we need to calculate the next `quorum_threshold` are the `total_staked_cycle` in the current stage.

The logic I'm proposing for `update_quorum` is:
- if we go from a `voting` stage to the `proposing` one following we do:
   1. update the `last_updated_stage` to the current one
   2. use `total_staked_cycle` to calculate and update the `quorum_threshold`
   3. set `total_staked_cycle` to `0`
- if we go from a `proposing` stage to the `voting` following (so in the same cycle) we do:
   1. update the `last_updated_stage` to the current one
   2. keep `total_staked_cycle` and `quorum_threshold` the same
 
Note that the advantage is that after doing this we can:
- in `vote` and `propose` just add to the `total_staked_cycle` (as we know it's for the current cycle)
- in `flush` use the `quorum_threshold`, as we know it's updated to the current stage/cycle already

Reply:
I saw now the rest of the comments about `vote` not needing this call, I think you are right in that we can store the `last_updated_cycle` instead of the `last_updated_stage`.

However the point above remains, I think the total staked in the cycle can (and should) be stored together.